### PR TITLE
Ensure local_used_maps_tmp is distinct from local_used_maps_[i]

### PR DESCRIPTION
**Pull request by mcarilli**
Followup/hotfix for https://github.com/pytorch/pytorch/pull/53160. @rohan-varma and @zhaojuanmao were seeing https://github.com/pytorch/pytorch/pull/53160/files#diff-9273e5ff7b40f30d6a4444d1c7be9fe9a5c2068070c68af4e7b0ac2d4cff0923R582 fire in some internal workloads, indicating `local_used_maps_tmp` wasn't actually being created as a distinct temporary, in other words, `local_used_maps_[i]` was already pinned for some reason. This seems like a bug with the CPU allocator: [`local_used_maps_` should not have been pinned on construction](https://github.com/pytorch/pytorch/blob/9be4c75fa0399a292e95ef8f3c79457e4b5b2338/torch/lib/c10d/reducer.cpp#L180-L183). We should [investigate that separately](https://github.com/pytorch/pytorch/pull/53160/files#r599188373).

In the meantime, the present PR should ensure `local_used_maps_tmp` is always distinct from `local_used_maps_[i]` (and therefore prevents the race condition described in https://github.com/pytorch/pytorch/pull/51360) even if `local_used_maps_[i]`is already pinned.

### CODE REVIEW

File: torch/lib/c10d/reducer.cpp [hunk 1]
```diff
@@ -577,9 +577,11 @@ void Reducer::mark_variable_ready(VariableIndex index) {
           //
           // ** In the hoped-for case where all params are used, DDP itself won't do any
           // blocking work between now and the re-zeroing, so the danger is real.
-          auto local_used_maps_tmp = local_used_maps_[i].pin_memory();
-          // Defensively ensures a deep copy to a pinned temporary
-          TORCH_INTERNAL_ASSERT(local_used_maps_tmp.data_ptr() != local_used_maps_[i].data_ptr())
```
Comment (zhaojuanmao) [inline torch/lib/c10d/reducer.cpp:582]: could we still keep the assert?

**Review by rohan-varma — COMMENTED** [review]
Thanks for the quick fix! We will test this on the errors we are seeing and circle back.

File: torch/lib/c10d/reducer.cpp (hunk 1 above)
Comment (mcarilli) [reply torch/lib/c10d/reducer.cpp:582]: Sure, better safe than sorry.

**Review by rohan-varma — APPROVED** [review]
Our internal tests are passing with this change, thanks for the quick fix!

**Pull request closed without merging**